analytics-core/src/main/java/com/segment/analytics/SegmentDispatcher.java

```diff
@@ -251,7 +251,12 @@ void performFlush() {
     try {
       int payloadsUploaded = networkExecutor.submit(new Callable<Integer>() {
         @Override public Integer call() throws Exception {
-          return upload();
+            try {
+                return upload();
+            } catch(IOException ioe) {
+                error(ioe, "Upload failed: " + ioe.getMessage());
+                return -1;
```

Comment:
Right, I guess my question is why aren't you just throwing the exception again instead of returning a number. The problem is that you're masking the exception by returning a number when you should be throwing an exception. The IOException is handled in the catch block here- https://github.com/kellyfj/analytics-android/blob/master/analytics-core/src/main/java/com/segment/analytics/SegmentDispatcher.java#L289.

It's partly queueFile's fault that that it doesn't throw an exception if you pass in a negative or 0 to remove. When you call `queueFile.remove(-1)`, it actually ends up adding a blank element. I'll fix that separately.

By return an invalid number instead of throwing an exception, the dispatcher thinks the upload completed successfully, then tries to remove a negative number of elements,  sees that are still elements left in the queue so it triggers another flush. Because of the blank element inserted by QueueFile, you'll start noticing that our servers reject those payloads, since they have a blank entry in an array, which is invalid JSON.

If you change this to return 0 (make sure you clear the application data first since the queue is corrupted from the previous runs) - you'll see that you won't see any of those rejected exceptions again - https://cloudup.com/cgCVjBAYS78.